### DISCUSSION TITLE
Can you publish multiple cameras using multiple rpiCamera sources?

### DISCUSSION THREAD

**Discussion by marksutheran** [bluenviron/mediamtx#1539] [Questions] [answered]
Is it possible to have multiple paths which are paired to multiple cameras on the RPi module (e.g. using CM4)?

Libcamera seems happy enough to run both simultaneously and using rtsp-simple-server I can switch the camera id to select each camera independently using only one path. However, if I attempt to create two paths each pointing to a different camera I get the following error:

  ERR: 'rpiCamera' is used as a source in multiple paths ('cam0' and 'cam1') 

Which appears to be by design looking @ path.go - but maybe I just don't understand the configuration?

As a workaround I can use a publisher endpoint and libcamera/ffmpeg to feed the second camera it but I was wondering if there is a cleaner way?

Thanks!

ps) This is an awesome media server!

**Comment by marksutheran** [comment]
Seems a sensible question given there are 2x CSI interfaces but can only use one at a time?

**Comment by aler9** [comment]
Hello, are you sure that libcamera is able to run two streams at the same time and V4L is able to encode two streams in H264 at the same time? if you confirm me that it's possible, then the server can be edited in order to support multiple CSI cameras, each with a different ID.

**Comment by marksutheran** [comment, marked as answer]
Hi Alessandro. Thanks for the reply... this really is a nice piece of software!

TLDR: Yes, I've been using this setup in a demo for almost a week

Details:
I don't know exactly *how* but libcamera can mux and encode both streams simultaneously even with two different processes. Putting ffmpeg in the middle is going to involve some extra copying but essentially it's stable.

Setup:
2x cams, 1920x1080 @ 20fps simultaneously

Exact setup config is shown below, kindly let me know if you would like to test other configurations.

> rtsp-simple-server.yaml

```
paths:
  cam0:
    source: publisher
  cam1:
    source: rpiCamera
    rpiCameraCamID: 0
    rpiCameraWidth: 1920
    rpiCameraHeight: 1080
    # etc...
```

> ffmpeg
`libcamera-vid --camera 1 --framerate 20 --nopreview --inline -t 0 --width 1920 --height 1080 --listen -o - | ffmpeg -i - -vcodec copy -f rtsp -rtsp_transport tcp rtsp://localhost:8554/cam0`

> representative top
```
    Tasks: 137 total,   1 running, 136 sleeping,   0 stopped,   0 zombie
    %Cpu(s):  6.4 us,  7.3 sy,  0.0 ni, 86.0 id,  0.0 wa,  0.0 hi,  0.3 si,  0.0 st
    MiB Mem :   7812.3 total,   7401.3 free,    231.1 used,    180.0 buff/cache
    MiB Swap:    100.0 total,    100.0 free,      0.0 used.   7472.4 avail Mem    
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
    962 mark      20   0  721488  33228  14088 S  31.5   0.4   0:59.76 rtsp-simple-ser
    970 mark      20   0  613188  12904  11196 S   9.1   0.2   0:35.13 libcamera-vid
    982 mark      20   0  350588  10832   9452 S   7.6   0.1   0:31.27 rtspss-embedded
     72 root       0 -20       0      0      0 I   6.6   0.0   0:02.58 kworker/u9:0-brcmf_wq/mmc1:0001:1
    971 mark      20   0  202832  47128  33704 S   3.6   0.6   0:14.90 ffmpeg
   1027 mark      20   0    9980   3240   2700 R   1.5   0.0   0:00.99 top
     93 root       1 -19       0      0      0 S   1.0   0.0   0:03.32 vchiq-slot/0
     15 root      20   0       0      0      0 I   0.5   0.0   0:01.38 rcu_preempt
    207 root      10 -10       0      0      0 S   0.5   0.0   0:01.84 SMIO
    783 root       0 -20       0      0      0 I   0.5   0.0   0:05.06 kworker/u9:2-brcmf_wq/mmc1:0001:1
   1008 root      20   0       0      0      0 I   0.5   0.0   0:00.35 kworker/3:3-events
   1029 root      20   0       0      0      0 I   0.5   0.0   0:00.50 kworker/0:1-events
      1 root      20   0  165224  10356   7404 S   0.0   0.1   0:03.53 systemd
```
      

I can view the streams simultaneously via either RTSP or WebRTC clients, e.g      

[image](https://ibb.co/vDVJtZC)

**Reply by aler9** [reply]
@marksutheran i've created a feature request here #1573 where i'll post a nightly release that allows using multiple RPI cameras at once. We can continue the discussion there.
